app/services/variants_stock_levels.rb

```diff
@@ -0,0 +1,38 @@
+# Report the stock levels of:
+#   - all variants in the order
+#   - all requested variant ids
+class VariantsStockLevels
+  def call(order, requested_variant_ids)
+    variant_stock_levels = variant_stock_levels(order.line_items)
+
+    # Potentially, the following lines are dead code, they are never reached
+    # Additionally, variants are not scoped here and so the stock levels reported would be incorrect
+    # See cart_controller_spec for more details and #3222
+    li_variant_ids = variant_stock_levels.keys
+    (requested_variant_ids - li_variant_ids).each do |variant_id|
```

Comment:
If I understand your comments correctly, you are saying that `requested_variant_ids - li_variant_ids` is always empty. That means that the inside of the `each` block is never executed. It does sound bizarre. There is one case I can think of:

- Apples have a stock level of 5.
- The user adds 5 apples to their cart.
- Someone else buys 5 apples at the same time
- The cart items are capped at stock levels and the apples are removed.
- The apples are not part of the order any more, but still in the request.

Sounds very unlikely, but may be possible. I also noticed that the cart service doesn't just look at `params[:variants]`, but also at `:products` and `:quantity`. Are there other interfaces sending requests with these ids?

Just playing around with the site I noticed an even more important bug: If an item is not available anymore, the server responds with 412, but the Javascript sees that as an error and enters an infinite loop of re-trying the request. The Update button stays on "Updating...". There should be an out-of-stock message instead.

Reply:
I am not sure about you scenario because cart_populate is after cap_quantity_at_stock so the item would still be in the order. and is cap_quantity_at_stock is not removing the item from the order? maybe some race condition between requests...

> Sounds very unlikely, but may be possible. I also noticed that the cart service doesn't just look at params[:variants], but also at :products and :quantity. Are there other interfaces sending requests with these ids?

Yes, I know. I cant find other calls to populate. The main one only sends variants.cart_service.read_products_hash . It's dead code isn't it? even the specs are not reaching it.
shall I remove it?

For that last part, we need to create an issue.

Reply:
I am tempted to add a bugsnag dead code detector on both the stock levels calculation and the :products hash. Shall I do that?

Reply:
I wish we had app monitoring already in place 😭 these go there, to Datadog.

Reply:
I am not sure how to use datadog for this... let me know if you think it's worth going through that now.

I have added bugsnag notifications, please see the new commit.